keps/sig-api-machinery/4222-cbor-serializer/README.md

```diff
@@ -613,7 +621,178 @@ sequence that it will not successfully decode.
+| 2  | static  | static  | request   | cbor     |
+| 3  | static  | static  | response  | json     |
+| 4  | static  | static  | response  | cbor     |
+| 5  | dynamic | static  | request   | json     |
+| 6  | dynamic | static  | request   | cbor     |
+| 7  | dynamic | static  | response  | json     |
+| 8  | dynamic | static  | response  | cbor     |
+| 9  | static  | dynamic | request   | json     |
+| 10 | static  | dynamic | request   | cbor     |
+| 11 | static  | dynamic | response  | json     |
+| 12 | static  | dynamic | response  | cbor     |
+| 13 | static  | static  | request   | protobuf |
+| 14 | static  | static  | response  | protobuf |
+
+In cases 5-6 and 11-12, the marshalling side acts on an Unstructured object and is not aware that
+the unmarshalling side may decode some of the payload into a `RawExtension`. The bytes stored in the
+`RawExtension` by unmarshalling ultimately depend on the negotiated content type, which can vary
+with the enablement of the CBOR serializer. Existing programs have so far been able to assume that
+unmarshalled RawExtensions always have either nil or a valid JSON text in their Raw field.
+
+In cases 1-4, 7-8, and 9-10, if the marshalling side populates Raw with a non-nil slice, it is
+responsible for ensuring that that encoding of the slice contents matches the encoding that will be
+used to serialize the object containing the `RawExtension`. This is trivially ensured in cases 13
+and 14 because Protobuf is capable of representing `RawExtension` values containing arbitrary
+bytes. Protobuf is not a supported encoding for Unstructured objects. Existing programs have in
+practice stored JSON in the Raw field of `RawExtension`.
+
+#### Compatibility
+
+If the `RawExtension` marshalling and unmarshalling behavior for CBOR were to be implemented in
+exactly the same way as the existing JSON behaviors, the assumptions in many existing programs that
+the Raw field can be assigned to a slice of JSON bytes, or that the Raw bytes of an unmarshalled
+`RawExtension` are valid JSON, would be broken.
+
+The simple approach of automatically transcoding JSON to CBOR during CBOR marshalling, and
+transcoding CBOR to JSON during CBOR unmarshalling, would avoid breaking existing programs. However,
+the expense of transcoding to or from JSON would negate any performance advantage of a binary
+encoding. This expense would not be limited to a few API types: significant examples include the use
+of a `RawExtension` field in `metav1.WatchEvent` to represent each watch event's object state, or
+the arbitrary objects embedded in `admissionv1.AdmissionRequest`.
+
+Instead, automatic transcoding to and from JSON will become optional and initially enabled by
+default.
+
+For marshalling:
+
+1. If both Raw and Object are `nil`, `null` is returned.
+1. If Object is not `nil`, return the result of marshalling Object.
+1. If Raw is prefixed by the CBOR encoding of the head of the "self-described CBOR" tag, then return
```

Comment:
is this proposal specifically for a RawExtension#MarshalCBOR implementation?

doesn't the reverse have to happen for RawExtension#MarshalJSON (e.g. looking for a CBOR prefix and transcoding CBOR → JSON)?

Reply:
Yes, you're right. Thanks for pointing it out, I will update this.

Reply:
should we augment RawExtension with an unserializable ContentType field to identify the content type, similar to runtime.Unknown, so that we don't have to rely on heuristics detecting prefixes of Raw data for instances of RawExtension?

RawExtension#UnmarshalJSON would set the field to ContentTypeJSON, RawExtension#UnmarshalCBOR would set the field to ContentTypeCBOR, so when re-marshaling into a different format we don't have to guess our content type

Reply:
This was discussed in https://github.com/kubernetes/kubernetes/pull/23112/files#r56546965 (leading to the current TODO comment) and I don't think I understand if there is a counterargument to doing this today. At least, doing what you have suggested sounds reasonable to me.

Reply:
Good catch, my mental model was that since there is a way to know if the Raw data is JSON or CBOR, we'd always trancode to the format the caller was asking for, but I don't see that specifically called out.

Reply:
Updated. IIUC, the content type would need to be serialized for Protobuf since it is not implied during unmarshaling as it would be for JSON or CBOR.

Reply:
interestingly, the conversion built-in types round-trip RawExtension through just blindly sets Unknown content type to JSON
 https://github.com/kubernetes/kubernetes/blob/ba28796e778731bdb8dfeaa9a85454ff28306c1a/staging/src/k8s.io/apimachinery/pkg/runtime/embedded.go#L118-L135

I guess all built-in persisted types should be validating the data in RawExtension is actually JSON before persisting today (and are not, currently)